flex_model/distributed/mappings.py

```diff
@@ -237,30 +238,301 @@ def scatter_data_parallel(tensor: Tensor, dim: int = 0) -> Tensor:
+            fp16_groups[name] = tensor
+        elif tensor.dtype == torch.bfloat16:
+            bf16_groups[name] = tensor
+        else:
+            raise NotImplementedError(
+                f"Tensor with dtype: {tensor.dtype} is not supported for "
+                f"gathering across PP ranks."
+            )
+
+    dtype_groups = {
+        torch.float32: fp32_groups,
+        torch.float16: fp16_groups,
+        torch.bfloat16: bf16_groups,
+    }
+
+    return dtype_groups
+
+
+# Tensor buffer metadata type.
+_TBUF_META = Dict[
+    str, Union[
+        int,
+        torch.dtype,
+        Dict[str, Tuple[int, int]],
+        Dict[str, torch.Size],
+    ]
+]
+
+
+def _make_flat_buffer(
+    tensor_dict: Dict[str, Tensor],
+) -> Tuple[Optional[Tensor], Optional[_TBUF_META]]:
+    tensors = []
+    name_to_index_map = {}
+    name_to_shape_map = {}
+    curr_idx = 0
+    for name, tensor in tensor_dict.items():
+        shape = tensor.shape
+        numel = tensor.numel()
+        tensors.append(tensor.flatten())
+
+        name_to_index_map[name] = (curr_idx, curr_idx + numel)
+        name_to_shape_map[name] = shape
+
+        curr_idx += numel
+
+    if len(tensors) == 0:
+        return None, None
+
+    tensor_buffer = torch.cat(tensors)
+
+    meta: _TBUF_META = {
+        "buffer_rank": dist.get_activation_pipeline_parallel_rank(),
+        "buffer_size": tensor_buffer.numel(),
+        "buffer_dtype": tensor_buffer.dtype,
+        "name_to_index_map": name_to_index_map,
+        "name_to_shape_map": name_to_shape_map,
+    }
+
+    return tensor_buffer, meta
+
+
+def _gather_pipeline_parallel(
+    tbuf_groups,
+    all_metadata_groups,
+) -> Dict[str, Tensor]:
+    world_size = dist.get_activation_pipeline_parallel_world_size()
+    rank = dist.get_activation_pipeline_parallel_rank()
+
+    # Setup collections for communication
+    def _empty_groups() -> Dict[torch.dtype, List[Union[Tensor, int]]]:
+        return {dtype: [] for dtype in tbuf_groups.keys()}
+
+    recv_tbuf_groups = _empty_groups()
+    recv_rank_groups = _empty_groups()
+    send_tbuf_groups = _empty_groups()
+    send_rank_groups = _empty_groups()
+
+    # Construct recv tensors and src ranks.
+    # NOTE: Only rank0 participates in recvs.
+    if rank == 0:
+        for metadata_groups in all_metadata_groups:
+            if metadata_groups is not None:
+                for dtype, metadata in metadata_groups.items():
+                    # Skip if there's no tbuf to recv for the dtype.
+                    if metadata is None:
+                        continue
+
+                    buffer_rank = metadata["buffer_rank"]
+                    buffer_size = metadata["buffer_size"]
+                    buffer_dtype = metadata["buffer_dtype"]
+                    assert buffer_dtype == dtype, (
+                        f"Dtype mismatch: {buffer_dtype} and {dtype}")
+
+                    # Skip if the buffer src is rank0.
+                    if buffer_rank == 0:
```

Comment:
Can collapse this with the `continue` statement above i.e. `if metadata["buffer_rank"] == 0`